compiler/rustc_codegen_llvm/src/context.rs

```diff
@@ -859,7 +859,11 @@ impl<'ll> CodegenCx<'ll, '_> {
 
         // This isn't an "LLVM intrinsic", but LLVM's optimization passes
         // recognize it like one and we assume it exists in `core::slice::cmp`
-        ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32);
+        match self.sess().target.arch.as_str() {
+            "avr" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
```

Comment:
```suggestion
            "avr" | "msp430" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
```
(And in `intrinsic.rs` too.)